File: solutions/csharp/grade-school/1/GradeSchool.cs
using System.Collections.Generic;
using System.Linq;

public class GradeSchool
{
    private readonly Dictionary<int, List<string>> _grades = new(); 
    
    public bool Add(string student, int grade)
    {
        if (!_grades.ContainsKey(grade))
        {
            _grades[grade] = new List<string>();
        }

        if(_grades.Values.Any(lst => lst.Contains(student)))
        {
            return false;
        }

        _grades[grade].Add(student);
        return true;
    }

    public IEnumerable<string> Grade(int grade)
    {
        if (!_grades.ContainsKey(grade))
        {
            return Enumerable.Empty<string>();
        }

        return _grades[grade].OrderBy(n => n);
    }

    public IEnumerable<string> Roster()
    {
        return _grades
            .OrderBy(g => g.Key)
            .SelectMany(g => g.Value.OrderBy(n => n));
    }
}

File: solutions/cpp/grade-school/1/grade_school.cpp
#include "grade_school.h"
#include <algorithm>

namespace grade_school {

    void school::add(const std::string& name, int grade) {
        grades[grade].push_back(name);
        std::sort(grades[grade].begin(), grades[grade].end());
    }
    
    std::vector<std::string> school::grade(int year) const {
        if(grades.count(year) == 0) {
            return {};
        }
        return grades.at(year);
    }
    
    std::map<int, std::vector<std::string>> school::roster() const {
        return grades;
    }

}  // namespace grade_school
